rpc/src/client/transport/websocket.rs

```diff
@@ -246,124 +340,146 @@ impl WebSocketClientDriver {
+        Ok(())
+    }
+
+    async fn publish_event(&mut self, ev: Event) {
+        if let PublishResult::AllDisconnected = self.router.publish(&ev) {
+            debug!(
+                "All subscribers for query \"{}\" have disconnected. Unsubscribing from query...",
+                ev.query
+            );
+            // If all subscribers have disconnected for this query, we need to
+            // unsubscribe from it. We issue a fire-and-forget unsubscribe
+            // message.
+            if let Err(e) = self
+                .send_request(Wrapper::new(unsubscribe::Request::new(ev.query.clone())))
+                .await
+            {
+                error!("Failed to send unsubscribe request: {}", e);
             }
         }
     }
 
-    async fn handle_generic_response(&mut self, wrapper: response::Wrapper<GenericJSONResponse>) {
-        let req_id = wrapper.id().to_string();
-        if let Some(state) = self.router.subscription_state(&req_id) {
-            match wrapper.into_result() {
-                Ok(_) => match state {
-                    SubscriptionState::Pending => {
-                        let _ = self.router.confirm_add(&req_id).await;
-                    }
-                    SubscriptionState::Cancelling => {
-                        let _ = self.router.confirm_remove(&req_id).await;
-                    }
-                    SubscriptionState::Active => {
-                        if let Some(event_tx) = self.router.get_active_subscription_mut(
-                            &SubscriptionId::from_str(&req_id).unwrap(),
-                        ) {
-                            let _ = event_tx.send(
-                                Err(Error::websocket_error(
-                                    "failed to parse incoming response from remote WebSocket endpoint - does this client support the remote's RPC version?",
-                                )),
-                            ).await;
-                        }
-                    }
-                },
-                Err(e) => match state {
-                    SubscriptionState::Pending => {
-                        let _ = self.router.cancel_add(&req_id, e).await;
-                    }
-                    SubscriptionState::Cancelling => {
-                        let _ = self.router.cancel_remove(&req_id, e).await;
-                    }
-                    // This is important to allow the remote endpoint to
-                    // arbitrarily send error responses back to specific
-                    // subscriptions.
-                    SubscriptionState::Active => {
-                        if let Some(event_tx) = self.router.get_active_subscription_mut(
-                            &SubscriptionId::from_str(&req_id).unwrap(),
-                        ) {
-                            let _ = event_tx.send(Err(e)).await;
-                        }
-                    }
-                },
+    async fn confirm_pending_command(
```

Comment:
confirm here means sending response?

Reply:
Yes. I can rename it to something a bit clearer though.